code review

bebebebebe opened the pull request:
This pull request is a comparison of master against a branch with only the initial commit. This is to facilitate code review, since github has inline commenting available for pull requests. To use, click on the "files changed" tab above, and comment away by hovering on the '+' to the left of lines. 

Inline comments can also be made on a commit level, so it's also possible to comment on the initial commit by looking at that commit individually.

Let me know if you run into issues, or questions about reviewing!

bebebebebe committed:
Client message constructors and send method

bebebebebe committed:
Server: parse incoming message, add connecting client to client list

bebebebebe committed:
Update readme

bebebebebe committed:
Server: send messages in outbox

bebebebebe committed:
Correct outgoing chat message format

bebebebebe committed:
Small refactor to shorten message format check method

bebebebebe committed:
Client: connect handshake, parse incoming messages, checking format/sender

bebebebebe committed:
Update readme

bebebebebe committed:
Client: monitor stdin and socket, sending stdin messages to server, and socket messages to stdout

bebebebebe committed:
Shorten tick length, so text moves faster to left

bebebebebe committed:
Regex check input before sending to server, so that user cannot effect eg a backspace or delete or return

bebebebebe committed:
Update readme

bebebebebe committed:
Add 'colors' key to each client, assigning integer from those left available of a pre set number of integers (default 5)

bebebebebe committed:
Oops, typo fixes

bebebebebe committed:
Send quit message to server when client quits

bebebebebe committed:
Change chat text representation to array instead of string, so can store data about speaker or text color

bebebebebe committed:
Different chatters have different coloured text

bebebebebe committed:
Client: disregard chat messages that are older than ones already received

bebebebebe committed:
Display list of client names, with current speaker name emphasized

bebebebebe committed:
Take optional commandline arg to specify server address

bebebebebe committed:
Update readme

bebebebebe committed:
Updates to client field check, readme to reflect updated server message formats

bebebebebe committed:
More readme updates

bebebebebe committed:
Readme: Trying to get fragment identifiers to work

bebebebebe committed:
More readme details

bebebebebe committed:
Move Console module to own file, rename files with name of class they contain

bebebebebe committed:
Readme: fix typos

tansengming commented on Console.rb:
```diff
@@ -0,0 +1,61 @@
+
+module Console
+	def self.color(color, text)
+		return text if color.nil?
```
it's a good idea to filter `color` because of the `send`. e.g.

```ruby
COLORS = %w{ red green blue }
def self.color(color, text)
  return text unless (color && COLORS.include?(color))
  # ...
```

tansengming commented on Console.rb:
```diff
@@ -0,0 +1,61 @@
+		self.public_send(color, text)
+	end
+
+  def self.red(text)
```
I recommend doing a little metaprogramming to clean up all these color methods, e.g.

```ruby
COLORS_WITH_CODES = {red: 31, green: 32}

COLORS_WITH_CODES.each do |color, code|
  define_singleton_method color do |text|
    color_encode(text, code)
  end
end

```

tansengming commented on InterruptClient.rb:
```diff
@@ -0,0 +1,231 @@
+		terminal_config
+		receive_loop
+	end
+
```
If I'm reading this correctly you might want to mark everything below this line as private methods.

tansengming commented on InterruptClient.rb:
```diff
@@ -0,0 +1,231 @@
+		begin
+			msg = JSON.parse(data)
+			msg = check_fields(msg)
+		rescue
```
a blanket rescue is generally considered an anti-pattern. Try to be a little more specific.

tansengming commented on InterruptClient.rb:
```diff
@@ -0,0 +1,231 @@
+	# return nil if fields wrong, otherwise return msg
+	def check_fields(msg)
+		basic_fields = (msg.has_key?('type') && msg.has_key?('time'))
+		return nil if not basic_fields
```
maybe it reads better `return nil if not basic_fields` ➡️  `return nil unless basic_fields` ?

jayschwa commented on Console.rb:
```diff
@@ -0,0 +1,61 @@
+	def self.color(color, text)
+		return text if color.nil?
+		self.public_send(color, text)
+	end
```
This method is indented more than other parts of the module. Is this intentional?

bebebebebe replied:
Good point. It looks fine in my version in sublime, but odd here, will have to see what's going on.

jayschwa replied:
Probably mixed whitespace. Sublime will show you tabs vs spaces when text is highlighted.

tansengming replied:
I've been burned by this so many times that I just enable whitespace glyphs all the time http://sublimetexttips.com/show-whitespace-sublime-text/

jayschwa commented on Console.rb:
```diff
@@ -0,0 +1,61 @@
+  end
+
+  def self.left(num)
+		"\033[#{num}D"
```
Use of `tput` would theoretically increase compatibilty. I don't have much Ruby experience, but it seems like use of backticks to execute a system command and get its output would be pretty convenient, and just as terse as what you're doing now.

The main downside is that calling another process for every keystroke will be "expensive" (but probably still imperceptible to the user). If that is an issue, `tput` could be called once at initialization and the output saved for later use in these methods.

jayschwa commented on InterruptClient.rb:
```diff
@@ -0,0 +1,231 @@
+			chat_array = msg['body']
+			names_array = msg['names']
+
+			chat_string = chat_string(chat_array)
```
Is this creating a new variable that shadows the function of the same name? I suggest using different names.

jayschwa commented on InterruptClient.rb:
```diff
@@ -0,0 +1,231 @@
+
+		names_array.each {|item|
+			nickname, color_code, emph = item
+			names_length += nickname.length + 2
```
What is the purpose of `names_length`? Unless I'm missing something, it looks identical to `names_string.length`.

bebebebebe replied:
The problem is that names_string contains extra characters for encoding color etc.

jayschwa commented on InterruptClient.rb:
```diff
@@ -0,0 +1,231 @@
+	FAREWELL = "\r\nbye"
+	PROMPT = '> '
+
+	MAX_MSG_LENGTH = 3000 # max length of incoming message read
```
This is a bit of a tangent, but datagrams of this size will almost certainly be fragmented into multiple packets and require reassembly if sent across the Internet. This should be handled transparently for you, but wanted to point it out if you weren't aware.

Further reading: https://en.wikipedia.org/wiki/Maximum_transmission_unit#MTUs_for_common_media

jayschwa commented on InterruptServer.rb:
```diff
@@ -0,0 +1,212 @@
+		@server.bind(host, port)
+
+		@clients = {}
+		@colors_available = [*0..num_colors-1]
```
Does the server actually know about colors? It doesn't seem like something that really needs to be baked into the protocol or server. If this is just an integer, it would be more aptly named as a client "id" or "number" or something. The notion of color would only exist on the client side. One could imagine an alternative client that uses some other way of differentiating clients instead of color.

bebebebebe replied:
That's a good point; perhaps I should rename it. It could be called "voice", or "position" for instance. From the point of view of the server, it is a finite list of items that can be assigned to at most one client at a time.

jayschwa commented on InterruptServer.rb:
```diff
@@ -0,0 +1,212 @@
+
+class InterruptServer
+
+	CHAT_LENGTH = 45 # length of chat text user can see at one time
```
Why 45? It seems a little low considering how wide a terminal can get.

bebebebebe replied:
It's just a parameter I'm starting with, and could be adapted depending on what the experience is like. Interesting that your intuition is that it is low. I was thinking of reducing it! The idea is to have something like voice conversation, where words don't stick around after they are uttered. I would ideally like the length to be a normal number of characters that one might take in at one moment while reading.

jayschwa commented on README.md:
```diff
@@ -1,41 +1,102 @@
-Message formats:
+# Interrupt Chat
```
Great README 💯

jayschwa commented on README.md:
```diff
@@ -1,41 +1,102 @@
+### Sent by server
 
-Sent by server:
+chat: `{'type' => 'chat', 'body' => (Array), 'names' => (Array) 'time' => (timestamp String)}`
```
I'm interested to see a concrete example of a `body`. Given two users, it might look like this?
```
user 0 says "foobar "
user 1 says "bazbam"
results in:
f0o0o0b1a1z1b0a0r0 0b1a1m1
```

bebebebebe committed:
Whitespace fix

bebebebebe committed:
Cleanup: rename var, remove unneeded line

bebebebebe committed:
Merge pull request #2 from bebebebebe/cleanup

Cleanup

zormit commented on README.md:
```diff
@@ -1,41 +1,102 @@
+
+Run the client in a third window in the same way to chat between the two windows running the client.
+
+If you're running the client, and the server is running elsewhere at '[server-ip-address-string]', run the client with a command line argment
```
small typo

zormit commented on InterruptClient.rb:
```diff
@@ -0,0 +1,230 @@
+  def receive_loop
+    loop do
+      to_read = IO.select([@client, STDIN]) # check what's ready to read: socket, or terminal input
+      to_read[0].each { |ios|
```
These two lines confuse me a bit. But I think that's the fault of the IO API.

zormit reviewed (commented):
Ok I got started ^^ but got to go now... I want to review more, this is very cool. I agree with Jay, the README is great.

bebebebebe committed:
Introduce convenient command line args for running on local network

bebebebebe committed:
Merge pull request #3 from bebebebebe/cmd-line-args

Introduce convenient command line args for running on local network

bebebebebe committed:
Use system calls in Console module to set or reset terminal options, and to clear screen when start chat

tansengming commented:
Apologies for the comments on old files. It looks like I forgot to submit the reviews I made a few days back.

tansengming commented on Console.rb:
```diff
@@ -0,0 +1,61 @@
+# Utility for writing to terminal, mostly colors and cursor movements.
```
by convention, filenames should be lower case and snaked cased https://github.com/bbatsov/ruby-style-guide#snake-case-files

tansengming commented on InterruptClient.rb:
```diff
@@ -0,0 +1,230 @@
+  MAX_MSG_LENGTH = 3000 # max length of incoming message read
+  HANDSHAKE_WAIT = 2 # number of seconds to wait for ack from server before resending
+
+  COLORS = [
```
If you're sure there won't be spaces in your color names I recommend `COLORS = %w{ green magenta etc }`

tansengming commented on InterruptClient.rb:
```diff
@@ -0,0 +1,230 @@
+    @client.send(json, 0)
+  end
+
+  def msg_chat(body)
```
whenever I find myself writing multiple methods that start with the same thing, e.g. `msg_` it smells like it's supposed to be another object. I recommend having `msg_connect`, `msg_chat`, `msg_quit` in a new object.

tansengming commented on InterruptClient.rb:
```diff
@@ -0,0 +1,230 @@
+    @name = STDIN.gets.chomp
+  end
+
+  def handshake
```
this seems a tad clearer

```ruby
def handshake
  loop do
    send_msg(msg_connect(@name))
    incoming = IO.select([@client], nil, nil, HANDSHAKE_WAIT)
    break if (incoming && ack?)
  end
end
```

tansengming replied:
might want to call it `handshake_loop` too.

tansengming commented on InterruptClient.rb:
```diff
@@ -0,0 +1,230 @@
+
+  def receive_loop
+    loop do
+      to_read = IO.select([@client, STDIN]) # check what's ready to read: socket, or terminal input
```
looks like you can do

```ruby
client_socket, stdin_socket = IO.select([@client, STDIN])
```
http://ruby-doc.org/core-2.0.0/IO.html#method-c-select

that should save you from having to figure out where the IO select came from

tansengming commented on InterruptClient.rb:
```diff
@@ -0,0 +1,230 @@
+      chat_array = msg['body']
+      names_array = msg['names']
+
+      names_string, names_length = names_data(names_array)
```
smells like it should be another object. e.g.

```ruby
names_data = NamesData(names_array)
# then later you can do
# names_data.length names_data.string etc
```

tansengming commented on InterruptClient.rb:
```diff
@@ -0,0 +1,230 @@
+  # disregard eg return, delete, backspace keys presses
+  # only pass on to server word, punctuation, or space characters
+  # TODO: strip out two chars following '\e'
+  def handle_key(input)
```
I recommend just trapping SIGINT . Then you can just exit by hitting ctrl-c https://gist.github.com/sauloperez/6592971

botwhytho committed:
Added Instructions to Run from Docker

If your dependencies change (like if you had ruby gems
or bindings to curses/tput,etc. you will have to modify
this a bit but available to help

bebebebebe committed:
Use tput for cursor position, and center chat text

tansengming commented on InterruptServer.rb:
```diff
@@ -0,0 +1,233 @@
+    @outbox = [] # array of messages, as hashes with sender info
+  end
+
+  def run
```
if there is only 1 public instance method on an object, the convention is to use `#call` . I don't know the exact story behind that, but I bet it's got something to do with being able to invoke it this way http://stackoverflow.com/questions/19189632/why-does-ruby-call-the-call-method-when-i-dont-supply-the-method-name

bebebebebe committed:
Merge pull request #5 from bebebebebe/tput

Tput

bebebebebe committed:
Merge pull request #4 from botwhytho/master

Added Instructions to Run from Docker

tansengming commented on InterruptServer.rb:
```diff
@@ -0,0 +1,233 @@
+  end
+
+  def handle_outbox
+    while not @outbox.empty?
```
`while @outbox.any?` to save on using a negative

tansengming commented on InterruptServer.rb:
```diff
@@ -0,0 +1,233 @@
+    port = sender[1]
+    host = sender[2]
+    key = port.to_s + host # host + port is unique per client
+
```
`_, port, host = sender` would achieve the same thing.

tansengming commented on InterruptServer.rb:
```diff
@@ -0,0 +1,233 @@
+  end
+
+  # return nil if fields wrong, otherwise return msg
+  def check_fields(msg)
```
this is also ripe to be an object. It feels like it wants to be something like `Message#valid?`.

tansengming commented:
some general comments:

- I ❤️  this, it's clearly written. I'm glad I got to learn about setting up protocols with UDP.
- ❤️  that you made console it's own module.
- Having said that I wish there were at least some tests for this. Minitest / RSpec is not hard to set up. I can help! (if you need it)
- IMHO the client is going to be a PITA to test because of all the side effects. You're going to have to structure it differently to be able to test it effectively.
- my one other complaint is that you should use more classes and objects. As my friend Nick says, be not afraid of classes in Ruby.

tansengming commented on InterruptServer.rb:
```diff
@@ -0,0 +1,233 @@
+    puts "running on host: #{host}"
+
+    @clients = {}
+    @buckets_available = [*0..num_buckets-1]
```
`[*0..num_buckets-1]` this is new to me. 👍

tansengming commented on InterruptServer.rb:
```diff
@@ -0,0 +1,233 @@
+    }
+  end
+
+  def delete_client(key)
```
Do you handle the case when a client times out? Or when a client disconnects without sending an exit packet?

bebebebebe committed:
Use ctrl-c as quit command, and trap term signal to gracefully quit if program killed

bebebebebe committed:
Simplify color encoding methods

bebebebebe committed:
Some name validation for user supplied names

bebebebebe committed:
Shorter notation for array of words

bebebebebe committed:
Merge pull request #6 from bebebebebe/console

Console

bebebebebe committed:
Correct typo

bebebebebe committed:
Rename files

bebebebebe committed:
Also use lower case filename for console

bebebebebe committed:
Shorten handshake code

bebebebebe committed:
Slightly clarify use of select

bebebebebe committed:
Specify private methods as private, use call for only public method

bebebebebe committed:
simplification

bebebebebe committed:
Hide cursor while printing to prevent flash effect

bebebebebe committed:
Formatting/whitespace

tansengming commented on Console.rb:
```diff
@@ -0,0 +1,51 @@
+  end
+
+  def self.term_width
+    height, width = STDIN.winsize
```
seems like this is sufficient?

```ruby
def self.term_width
   STDIN.winsize[1]
end
```

tansengming commented on InterruptClient.rb:
```diff
@@ -0,0 +1,238 @@
+  # disregard eg return, delete, backspace keys presses
+  # only pass on to server word, punctuation, or space characters
+  def handle_key(input)
+    if input == CMD_QUIT
```
does your code ever get here? I'm quite sure that SIGTERM would take precedence over this. (I could be wrong!)

bebebebebe replied:
The terminal is in "raw" mode, so as to be able to capture keypresses before the user presses return. This means that ctrl-c won't send an interrupt/terminate signal.

tansengming replied:
I didn't know about raw mode thanks!

tansengming commented on InterruptClient.rb:
```diff
@@ -0,0 +1,238 @@
+  # return parsed msg if correct format and comes from Interrupt server
+  # return nil otherwise
+  def parse_msg(data, sender)
+    return nil if not is_server?(sender)
```
`return unless is_server?(sender)` is a bit clearer, although if you could do `return unless sender.is_server?` then it would be English!

bebebebebe committed:
Draw line under chat text plus some cleanup

bebebebebe committed:
Error handling to gracefully quit if server connection is refused.

bebebebebe committed:
Merge pull request #7 from bebebebebe/cleanup

Cleanup

bebebebebe committed:
Bugfix: emph text escape sequence was missing renormalize at end of string

bebebebebe committed:
Add thread to ping clients at intervals, delete ones that seem to have disappeared for a while

bebebebebe committed:
Add more colors

bebebebebe committed:
Cycle through colors again when run out, instead of leaving clients uncolored

bebebebebe committed:
Simplify the client monitor thread: all it does is sleep and periodically set an instance boolean variable to true, so the main thread can do monitoring activities and reset the boolean to false.

bebebebebe committed:
Merge pull request #8 from bebebebebe/monitor-clients

Monitor clients

bebebebebe committed:
Add instructions on readme to run server on network

bebebebebe committed:
Update readme with progress on limitations/todos

bebebebebe committed:
readme formatting

bebebebebe committed:
Give way to select from available ip addresses to run server from

bebebebebe committed:
Merge pull request #9 from bebebebebe/select-ip

Give way to select from available ip addresses to run server from

bebebebebe committed:
Update readme with info on new -network -select flag

bebebebebe committed:
Add video clip of chat to readme

bebebebebe committed:
Trying out adding gif to readme via github issues

bebebebebe committed:
Add link to presentation slides to readme

bebebebebe committed:
Update readme with info about client monitoring thread

bebebebebe committed:
Update readme with correct quit command
